lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                _contextManager.RemoveFailedContextInFailedEvaluator(failedEvaluator);
+
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.WaitingForEvaluator:
+                        if (!_evaluatorManager.ReachedMaximumNumberOfEvaluatorFailures())
+                        {
+                            if (isMaster)
+                            {
+                                Logger.Log(Level.Info, "Requesting a master Evaluator.");
+                                _evaluatorManager.RequestUpdateEvaluator();
+                            }
+                            else
+                            {
+                                _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                                Logger.Log(Level.Info, "Requesting mapper Evaluators.");
+                                _evaluatorManager.RequestMapEvaluators(1);
+                            }
+                        }
+                        else
+                        {
+                            Logger.Log(Level.Warning, "The system is not recoverable, change the state to Fail.");
+                            _systemState.MoveNext(SystemStateEvent.NotRecoverable);
+                            FailAction();
+                        }
+                        break;
+
+                    case SystemState.SubmittingTasks:
+                    case SystemState.TasksRunning:
+                        // set system state to ShuttingDown
+                        _systemState.MoveNext(SystemStateEvent.FailedNode);
+                        _taskManager.RecordTaskFailWhenReceivingFailedEvaluator(failedEvaluator);
+                        _taskManager.CloseAllRunningTasks(TaskManager.CloseTaskByDriver);
+
+                        // Push evaluator id back to PartitionIdProvider if it is not master
+                        if (!isMaster)
+                        {
+                            _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                        }
+
+                        TryRecovery();
+                        break;
+
+                    case SystemState.ShuttingDown:
+                        _taskManager.RecordTaskFailWhenReceivingFailedEvaluator(failedEvaluator);
+
+                        // Push evaluator id back to PartitionIdProvider if it is not master
+                        if (!isMaster)
+                        {
+                            _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                        }
+                        TryRecovery();
+                        break;
+
+                    case SystemState.Fail:
+                        Logger.Log(Level.Error, "Evaluator with Id: {0} failed System is in Fail state.", failedEvaluator.Id);
```

Comment:
not sure why this logging error message, while other logs are informational.
it might make sense to have this log globally (at the beginning of the handler) to create error record with current system state.